File: Vol.01(1001-1099)/#1084 Channel Allocation/p1084.cpp
#include <iostream> 
#include <string> 
using namespace std;

int main() {
    
    int       N,    max,    color[26];
    
    string    adj[26]; 
    
    while( cin >> N ) {
    
        if( !N ) break; 
    
        for( int i = 0; i < N; i++ )    cin >> adj[i];
    
        for( int i = 0; i < N; i++ )    color[i] = 0;
    
        max = 1; 
    
        for( int i = 0; i < N; i++ ) {
    
            bool    clr[26];
    
            for( int j = 0; j < N; j++ )    clr[j] = false; 
    
            for( int j = 2; j < adj[i].size(); j++ )
    
                if( color[adj[i][j] - 'A'] ) 
    
                    clr[color[adj[i][j] - 'A'] - 1] = true;
            int    x;
    
            for( x = 0; clr[x]; x++ );

            color[i] = x + 1;

            if( color[i] > max )    max = color[i]; 

        }

        if( max == 1 )    cout << "1 channel needed." << endl;
        else        cout << max << " channels needed." << endl; 

    }

} 

File: 范例/Accpt/10XX/p1084.pas
PROGRAM p1084;

VAR
        n,i                             :Integer;
        ch1,ch2                         :Char;
        edge                            :array[1..26,0..26]of Integer;
        whether,col                     :array[1..26,1..26]of Boolean;
        chose                           :array[1..26]of Boolean;
        ans,total                       :array[1..26]of Integer;


PROCEDURE Main;
var
        i,j,k,max,answer                :Integer;
begin
for i:=1 to n do
        for J:=1 to n do
                if whether[i,j] then
                        begin
                        inc(edge[i,0]);
                        edge[i,edge[i,0]]:=j;
                        end;


answer:=0;

for i:=1 to n do
        begin
        max:=-1;
        for J:=1 to n do
                if (ans[j]=0)and(total[j]>max) then
                        begin
                        max:=total[j];
                        k:=j;
                        end;
        for j:=1 to n do
                if not col[k,j] then break;
        ans[k]:=j;
        if not chose[ans[k]] then
                begin
                chose[ans[k]]:=true;
                inc(answer);
                end;
        for j:=1 to edge[k,0] do
                if (ans[edge[k,j]]=0)and(not col[edge[k,j],ans[k]]) then
                        begin
                        inc(total[j]);
                        col[edge[k,j],ans[k]]:=true;
                        end;
        end;



if answer=1 then writeln(answer,' channel needed.') else
        writeln(answer,' channels needed.');
end;

BEGIN
readln(n);
while (n<>0) do
        begin
        fillchar(edge,sizeof(edge),0);
        fillchar(whether,sizeof(whether),0);
        fillchar(col,sizeof(col),0);
        fillchar(ans,sizeof(ans),0);
        fillchar(total,sizeof(total),0);
        fillchar(chose,sizeof(chose),0);
        for i:=1 to n do
                begin
                read(ch1,ch2);
                while not eoln do
                        begin
                        read(ch2);
                        whether[ord(ch1)-ord('A')+1,ord(ch2)-ord('A')+1]:=true;
                        whether[ord(ch2)-ord('A')+1,ord(ch1)-ord('A')+1]:=true;
                        end;
                readln;
                end;
        Main;
        readln(n);
        end;
END.
